pkg/hub/submarinerbroker/controller.go

```diff
@@ -174,6 +174,11 @@ func (c *submarinerBrokerController) reconcileManagedClusterSet(ctx context.Cont
 		return c.doClusterSetCleanup(ctx, clusterSet, recorder)
 	}
 
+	if meta.IsStatusConditionPresentAndEqual(clusterSet.Status.Conditions, clusterv1beta2.ManagedClusterSetConditionEmpty,
```

Comment:
Does this API (reconcileManagedClusterSet) get invoked periodically? 

If so, can we think of a way to optimize this check as otherwise the condition (clusterv1beta2.ManagedClusterSetConditionEmpty) would always be true for an empty clusterSet and we will be invoking deleteBrokerResources() at regular intervals.

Reply:
It should only get invoked when something changes in the `ManagedClusterSet`.

Reply:
Okay, then its good.

Reply:
Yes, this is invoked periodically. That is why I have an extra check for finalizer. and didn't add any log here. There is also extra check in `deleteBrokerResources` that checks of broker namespace is present or not. Ideally finalizer+empty check should be enough but since we add finalizer even for clustersets without submariner, it doesn't help much yet.

Reply:
Ack. Can you point me to the code where we create the brokerNamespace only when a ManagedCluster is added to the clusterSet?

Reply:
@sridhargaddam We don't, ie we create the brokerNamespace when a ManagedClusterSet is created.  But only ManagedClusterSets of type ExclusiveClusterSetLabel.

Reply:
@vthapar What periodic update cadence are you seeing? I ran the integration tests and, after steady state, saw a `ManagedClusterSet` update after 10 min.

So the extra `Get` of the broker namespace in `deleteBrokerResources` is intended as a shortcut to reduce API calls due to periodic update. `DeleteFromManifests` does a single `Delete` but it attempts to delete 2 resources so the extra `Get` does save one API call.

But there's another simple check we could do. We combine adding the finalizer and the namespace label in [one Update operation](https://github.com/stolostron/submariner-addon/blob/a5cd8c42a701b16edd7566206cf9328d9c4d6d32/pkg/hub/submarinerbroker/controller.go#L179). But if we add the finalizer first and return when added, then we can also check  for the presence of the label:

```
if meta.IsStatusConditionPresentAndEqual(...) && finalizer.IsPresent(...) && 
            clusterSet.GetAnnotations()[SubmBrokerNamespaceKey] != "" {
    ....
}
```

Reply:
@tpantelis Sorry, I was not clear enough. You're right, it is not periodic but on clusteset chanes. But with this code if I only check for finalizer and empty to clean up, it was hitting repeatedly for `default` clusterset. For some reason we add finalizer even to default clusterset even when there is no managedcluster added to it, and that is what was causing periodic updates till I added check for namespace in `deleteBrokerResources`

Not sure I get comment about returning after adding finalizer. I am doing get on Namespace coz for `default` clusterset which had no clusters or submariner installed, I didn't see the namespace though the annotation and finalizer still exists. I added the Namespace check in called method so that in future when we only add finalizer to clustersets which have submariner installed, we can skip namespace Get in `deleteBrokerResources` and just use finalizer as required check.

Reply:
I do see a periodic update cadence of 10 min.

I don't understand why we get repeated updates w/o the namespace check. `DeleteFromManifests` just does a delete on the 2 broker resources. It doesn't make sense why that would cause repeated updates for the MCS. We need to understand that. I assume you had not added any `ManagedCluster` to the default MCS. I'll pull in your PR and try it with the integration tests.

When we add the finalizer and/or namespace annotation, if either was actually added we return immediately and don't apply the broker resources yet. We wait for the next update to do so. This has the side effect such that with your changes, we don't apply the broker resources if the MCS is empty. By separating adding the finalizer from the namespace annotation, I think we can check for presence of the namespace annotation instead of the namespace resource.

We really need unit tests to ensure correctness. I can pull in your PR and add tests.

Reply:
> due to deleting resources on default clusterset [including delete finalizer]

But you're not deleting the finalizer in  your new code.

> And won't adding a finalizer to clusterset trigger an event for subsequent update operation?

yes